File: nqueen.py
import random

def print_state(state):
    n = len(state)
    print("Fitness:", fitness(state))
    #print(state)
    for i in range(0,n):
        y = n-1 - i
        print(y,end="")
        for x in range(n):
            if state[x] == y:
                print('👸', end="")
                continue
            if x+y % 2 == 0 or (y+x) % 2 == 0:
                print('⬛', end="")
            else:
                print("⬜", end="")
                
        print("")
    print(" ",end="")
    for x in range(n):
        print(x, end=" ")
    print("")


def random_state(n: int):
    state = list(range(n))
    random.shuffle(state)
    return state

def populate(n: int, pop_len: int):
    population = []
    for i in range(pop_len):
        population.append(random_state(n))

    return population

def fitness(state):
    # FIXME BUG This only checks the diagonal over us not under
    # Here we sum the number of queens on the same diagonal
    n = len(state)
    diag_sum = 0
    for i in range(0,n):
        col = state[i]
        # here we check all the squares
        # diagonaly to our cell
        for j in range(i+1,n):
            # if the row above us 
            if abs(state[j] - state[i]) == abs(j - i):
                diag_sum += 1
    return diag_sum

def selection(population: list, amnt: int) -> list:
    population.sort(key=fitness)
    return population[:amnt]
    
def crossover(parents: list, mutation_rate: int = 0.05) -> list:
    # here we combine 2 parents into one child
    # we take one half and fill in the rest
    # from the other parent
    n = len(parents[0])
    half = (n+1) // 2

    child = parents[0][:half]
    for v in parents[1]:
        if v not in child:
            child.append(v)

    if random.random() < mutation_rate:
        i, j = random.sample(range(len(child)), 2)
        child[i], child[j] = child[j], child[i]

    return child

# ===== main program =====
def nqueen(n: int,
           pop_len: int,
           amnt_parents: int = 2,
           mutation_rate: int = 0.05) -> int:
    # solutions only exists for n >= 4
    assert(n >= 4)
    # to be able to select amnt_parents, population needs to be bigger
    assert(pop_len >= amnt_parents)
    population = populate(n,pop_len)
    found = False
    # play it out over generations
    generations = 10000
    for i in range(generations):
        parents = selection(population, amnt_parents)
        population = []
        for j in range(pop_len):
            child = crossover(parents, mutation_rate)
            population.append(child)
            if fitness(child) == 0:
                found = True
                break
        if found:
            return (i, child)
    return (-1, {})

if __name__ == "__main__":
    nqueen(8,10)

File: nqueen.c
#include <assert.h>
#include <math.h>
#include <stdbool.h>
#include <stdlib.h>
#include <stdio.h>
#include <strings.h>
#include <time.h>
#include <pthread.h>

#include "./nqueen.h"


Chromosome get_rand_r_state(size_t n, unsigned int* seed) {
  Chromosome state;
  state.pos = malloc(sizeof(int) * n);
  state.size = n;
  
  for (int i = 0; i < n; i++) {
    state.pos[i] = i;
  }

  for (int i = n - 1; i > 0; i--) {
    int j = rand_r(seed) % (i + 1);

    int tmp = state.pos[i];
    state.pos[i] = state.pos[j];
    state.pos[j] = tmp;
  }
  return state;
}

void populate(int n, Chromosome *population, size_t size, unsigned int* seed) {
  for (int i = 0; i < size ; i ++ ) {
    population[i] = get_rand_r_state(n, seed);
  }
}

int fitness(Chromosome state) {
  int diag_sum = 0;
  for (int i = 0; i < state.size; i ++) {
    for (int j = i+1; j < state.size; j ++) {
      if (abs(state.pos[j] -state.pos[i]) == abs(j - i) ) {
        diag_sum += 1;
      }
    }
  }
  return state.size-diag_sum;
}

int compare_fit(const void *a, const void *b) {
  const Chromosome *ca = a;
  const Chromosome *cb = b;
  
  return cb->fitness - ca->fitness;
}

void print_chromosome(Chromosome state) { 
  for(int i = 0; i < state.size; i ++) {
    printf("%d,", state.pos[i]);
  }
  printf("\nfitness %d\n", fitness(state));
}


void selection(Chromosome* population, size_t pop_len, Chromosome* parents, int amnt_parent) {
  qsort(population, pop_len, sizeof(Chromosome), compare_fit);

  for (int i = 0; i < amnt_parent; i ++ ) {
    parents[i].pos = malloc(sizeof(int) * population[i].size);
    for (int j = 0; j < population[i].size; j ++ ) {
      parents[i].pos[j] = population[i].pos[j];
    }
    parents[i].size = population[i].size;
  }
  
 }

void mutate(Chromosome *child, unsigned int* seed) {
  int i = rand_r(seed) % child->size;
  int j = rand_r(seed) % child->size;

  while (j == i)
    j = rand_r(seed) % child->size;

  int tmp = child->pos[i];
  child->pos[i] = child->pos[j];
  child->pos[j] = tmp;
}

// 2 parents
Chromosome crossover(Chromosome parent1, Chromosome parent2, float mutation_rate, unsigned int* seed) {
  // randommize the parent size not using half allways
  Chromosome child = {0};
  size_t n = parent1.size;
  child.pos = malloc(sizeof(int) * n);
  child.size = n;
  // copy the first half

  // too keep track of used values
  bool used[n];
  for (int i = 0; i < n; i ++){
    used[i] = false;
  }

  // copy the first half from parent 1x
  int pos = 0;
  for (int i = 0; i < n/2; i ++) {
    child.pos[i] = parent1.pos[i];
    used[parent1.pos[i]] = true;
    pos += 1;
  }

  // fill in the rest from parent 2
  for (int i = 0; i < n; i ++) {
    int value = parent2.pos[i];
    if(used[value] != true) {
      child.pos[pos] = value;
      pos += 1;
    }
  }
  // mutate if needed
  float ran = (float)rand_r(seed) / (float)RAND_MAX;
  if (ran < mutation_rate) {
    mutate(&child, seed);
  }
  
  return child;
}



int nqueen(Config config) {
  assert(config.pop_len >= config.amnt_parents);
  assert(config.n >= 4);
  assert(config.amnt_parents >= 2);

  // create the initial population
  Chromosome* population = malloc(sizeof(Chromosome)*config.pop_len);
  populate(config.n, population, config.pop_len, config.random_seed);
  for (int i = 0; i < config.pop_len; i ++) {
    population[i].fitness = fitness(population[i]);
  }

  // as default it is the max generations
  int sum = MAX_GENERATIONS;

  for (int i = 0; i < MAX_GENERATIONS; i ++) {
    Chromosome* parents = malloc(sizeof(Chromosome) * config.amnt_parents);
    selection(population, config.pop_len, parents, config.amnt_parents);
    
    Chromosome child = {0};
    for (int j = 0; j < config.pop_len; j ++) {

      // pick randomly the parents from the parents array
      int index1 = rand_r(config.random_seed) % config.amnt_parents;
      int index2 = rand_r(config.random_seed) % config.amnt_parents;
      while (index2 == index1){
        index2 = rand_r(config.random_seed) % config.amnt_parents;
      }

      child = crossover(parents[0], parents[1], config.mutation_rate, config.random_seed);
      
      // free populations
      free(population[j].pos);
      child.fitness = fitness(child);
      population[j] = child;

      // check child fitness (if it's complete we stop)
      if (child.fitness == config.n) {
        sum = i;
        for (int j = 0; j < config.amnt_parents; j ++ ) {
          free(parents[j].pos);
        }
        free(parents);
        goto DONE;
      }
    }
    for (int j = 0; j < config.amnt_parents; j ++ ) {
      free(parents[j].pos);
    }
    free(parents);
  }

 // free the population and return the sum
 DONE:
  for (int j = 0; j < config.pop_len; j ++ ) {
      free(population[j].pos);
    }
    free(population);
  
  return sum;
}

  
